react/__tests__/index.test.tsx

```diff
@@ -1,5 +1,160 @@
+
+    fireEvent.change(mockedInput, { target: { value: wrongEmail } })
+    fireEvent.click(submit)
+
+    expect(mockedInput).toHaveAttribute('data-errormessage')
+  })
+
+  it('should show error message if name is empty', () => {
+    const { getByPlaceholderText, getByText } = render(
+      <Newsletter>
+        <FormNameInput />
+        <FormSubmit />
+      </Newsletter>
+    )
+
+    const mockedInput = getByPlaceholderText(NAME_PLACEHOLDER_MESSAGE_ID)
+
+    const submit = getByText('Subscribe')
+
+    fireEvent.click(submit)
+
+    expect(mockedInput).toHaveAttribute('data-errormessage')
+  })
+})
+
+describe('Reacting to mutation results', () => {
+  it('should show success message if everything went well', async () => {
+    const validEmail = 'foobar@vtex.com'
+    const validName = 'Ted'
+
+    const mocks = [
+      {
+        request: {
+          query: subscribeNewsletter,
+          variables: { email: validEmail, name: validName },
+        },
+        result: {
+          data: {
+            subscribeNewsletter: {},
+          },
+        },
+      },
+    ]
+
+    const { getByPlaceholderText, getByText } = render(
+      <Newsletter>
+        <FormEmailInput />
+        <FormNameInput />
+        <FormSubmit />
+      </Newsletter>,
+      {
+        graphql: { mocks, addTypename: false },
+      }
+    )
+
+    const mockedEmailInput = getByPlaceholderText(EMAIL_PLACEHOLDER_MESSAGE_ID)
+    const mockedNameInput = getByPlaceholderText(NAME_PLACEHOLDER_MESSAGE_ID)
+
+    const submit = getByText('Subscribe')
+
+    fireEvent.change(mockedEmailInput, { target: { value: validEmail } })
+    fireEvent.change(mockedNameInput, { target: { value: validName } })
+
+    fireEvent.click(submit)
+
+    const thanks = await waitFor(() => getByText('Thanks for subscribing!'))
+
+    expect(thanks).toBeTruthy()
+  })
+
+  it('should show error message if something went wrong', async () => {
+    const validEmail = 'foobar@vtex.com'
+    const validName = 'Ted'
+
+    const mocks = [
+      {
+        request: {
+          query: subscribeNewsletter,
+          variables: { email: validEmail, name: validName },
+        },
+        error: new Error('Something went wrong'),
+      },
+    ]
+
+    const { getByPlaceholderText, getByText } = render(
+      <Newsletter>
+        <FormEmailInput />
+        <FormNameInput />
+        <FormSubmit />
+      </Newsletter>,
+      {
+        graphql: { mocks, addTypename: false },
+      }
+    )
+
+    const mockedEmailInput = getByPlaceholderText(EMAIL_PLACEHOLDER_MESSAGE_ID)
+    const mockedNameInput = getByPlaceholderText(NAME_PLACEHOLDER_MESSAGE_ID)
+
+    const submit = getByText('Subscribe')
+
+    fireEvent.change(mockedEmailInput, { target: { value: validEmail } })
+    fireEvent.change(mockedNameInput, { target: { value: validName } })
+
+    fireEvent.click(submit)
+
+    const sorry = await waitFor(() => getByText('Sorry, something went wrong.'))
+
+    expect(sorry).toBeTruthy()
```

Comment:
😍 😍 😍 😍 😍 😍 😍 😍 😍 😍 😍 😍

Reply:
This implementation actually comes from you, do you remember it?? 😂

Reply:
waaat noo hahahaha